scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala

```diff
@@ -976,15 +988,28 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
         case tpe: Type.Singleton => tpe
         case Type.Apply(underlying, args) => atPos(tpe, tpe)(Pat.Type.Apply(loop(underlying), args.map(convert)))
         case Type.ApplyInfix(lhs, op, rhs) => atPos(tpe, tpe)(Pat.Type.ApplyInfix(loop(lhs), op, loop(rhs)))
-        case Type.Function(params, res) => atPos(tpe, tpe)(Pat.Type.Function(params.map(p => loop(p.require[Type])), loop(res)))
+        case Type.Function(params, res) => atPos(tpe, tpe)(Pat.Type.Function(params.map {
+          case q @ Type.Arg.Quasi(1, Type.Arg.Quasi(0, _)) => q.become[Pat.Type.Quasi]
+          case p => loop(p.require[Type])
+        }, loop(res)))
         case Type.Tuple(elements) => atPos(tpe, tpe)(Pat.Type.Tuple(elements.map(loop)))
         case Type.Compound(tpes, refinement) => atPos(tpe, tpe)(Pat.Type.Compound(tpes.map(loop), refinement))
         case Type.Existential(underlying, quants) => atPos(tpe, tpe)(Pat.Type.Existential(loop(underlying), quants))
         case Type.Annotate(underlying, annots) => atPos(tpe, tpe)(Pat.Type.Annotate(loop(underlying), annots))
+        case Type.Placeholder(Type.Bounds(None, None)) => convert(tpe)
```

Comment:
patternTyp() called on `_` should give an error, because stuff like `parse 'x match { case y: _ => }'` produces an error. It's impossible to have a Pat.Var.Type at the top level of patternTyp (only as a type argument), and in the same vein it's impossible to have Pat.Type.Wildcard at the top level.